com.unity.shadergraph/Editor/Drawing/Controllers/BlackboardController.cs

```diff
@@ -1,21 +1,478 @@
+        {
+            Assert.IsNotNull(graphData, "GraphData is null while carrying out CopyShaderInputAction");
+            Assert.IsNotNull(shaderInputToCopy, "ShaderInputToCopy is null while carrying out CopyShaderInputAction");
+            switch (shaderInputToCopy)
+            {
+                case AbstractShaderProperty property:
+                    var copiedProperty = (AbstractShaderProperty)graphData.AddCopyOfShaderInput(property, insertIndex);
+                    if (copiedProperty != null) // some property types cannot be duplicated (unknown types)
+                    {
+                        // Update the property nodes that depends on the copied node
+                        foreach (var node in dependentNodeList)
+                        {
+                            if (node is PropertyNode propertyNode)
+                            {
+                                propertyNode.owner = graphData;
+                                propertyNode.property = copiedProperty;
+                            }
+                        }
+                    }
+
+                    copiedShaderInput = copiedProperty;
+                    break;
+
+                case ShaderKeyword shaderKeyword:
+                    // Don't duplicate built-in keywords within the same graph
+                    if (shaderKeyword.isBuiltIn && graphData.keywords.Any(p => p.referenceName == shaderInputToCopy.referenceName))
+                        return;
+
+                    var copiedKeyword = (ShaderKeyword)graphData.AddCopyOfShaderInput(shaderKeyword, insertIndex);
+
+                    // Update the keyword nodes that depends on the copied node
+                    foreach (var node in dependentNodeList)
+                    {
+                        if (node is KeywordNode propertyNode)
+                        {
+                            propertyNode.owner = graphData;
+                            propertyNode.keyword = copiedKeyword;
+                        }
+                    }
+
+                    copiedShaderInput = copiedKeyword;
+                    break;
+
+                default:
+                    Debug.Log("ERROR: BlackboardController: Unable to complete Copy Shader Input action.");
+                    return;
+            }
+        }
+
+        public Action<GraphData> modifyGraphDataAction => CopyShaderInput;
+
+        public IEnumerable<AbstractMaterialNode> dependentNodeList { get; set; } = new List<AbstractMaterialNode>();
+
+        public BlackboardItem shaderInputToCopy { get; set; }
+
+        public BlackboardItem copiedShaderInput { get; set; }
+
+        public int insertIndex { get; set; } = -1;
+    }
+
+    class BlackboardController : SGViewController<GraphData, BlackboardViewModel>
+    {
+        // Type changes (adds/removes of Types) only happen after a full assmebly reload so its safe to make this stuff static (I think, confirm this assumption)
```

Comment:
This should be true until we add custom structs - then we'll need to be more dynamic about it.